File: 2015/python/4/main.py
# Day 4: The Ideal Stocking Stuffer
# https://adventofcode.com/2015/day/4

from hashlib import md5
import os
import sys

def main():
    # get command line arguments
    if len(sys.argv) != 2:
        print("Usage: python main.py <input.txt>")
        sys.exit(1)

    file_name = sys.argv[1]

    # read input file
    prefix = ""
    with open(file_name, "r") as file:
        prefix = file.readline().strip()

    i = 0
    leading_5_found = False
    while True:
        input_val = prefix + str(i)
        hash_val = md5(input_val.encode()).hexdigest()
        
        if hash_val[:6] == "000000":
            print("Part 2:", i)
            break
        
        if hash_val[:5] == "00000" and not leading_5_found:
            print("Part 1:", i)
            leading_5_found = True

        i+=1

if __name__ == "__main__":
    main()

File: 2015/ruby/4/main.rb
#!/usr/bin/env ruby

require 'digest'

def main()
    args = ARGV

    if args.length < 1
        puts "Usage: ruby main.rb <input.txt>"
        exit 1
    end

    filename = args[0]
    file = File.open(filename, "r")

    if file.nil?
        puts "Could not open file"
        exit 1
    end

    prefix = file.readlines
    file.close

    leading_5_found = false
    i = 0
    while true
        input = prefix[0].strip + i.to_s
        hash = Digest::MD5.hexdigest(input)

        if hash[0..5] == "000000"
            puts "Part 2: #{i}"
            break
        end

        if hash[0..4] == "00000" && !leading_5_found
            puts "Part 1: #{i}"
            leading_5_found = true
        end

        i += 1
    end

end

main()